Refactor Filters for Text, Select and DateRange

galelis opened the pull request:
**OHIF-46
OHIF-47
OHIF-48**

### Descrioption
Separate into small components the filters for Text, Select and DateRage

### Screenshot
NOTE: Nothing should change on current layout. 
<img width="1217" alt="Screen Shot 2020-03-25 at 05 26 56" src="https://user-images.githubusercontent.com/2378326/77516478-44954780-6e59-11ea-8a53-0ef54337618c.png">

dannyrb commented:
If possible, modality should be:

- Multi-select
- Values should be all caps
- Checkboxes in expanded dropdown to add to list
- When collapsed, show values not as pills, but as text
- When collapsed, if value exceeds container, truncate with `...`

dannyrb commented on platform/ui/index.js:
```diff
@@ -10,6 +10,10 @@ export {
   DateRange,
+  FilterDateRange,
+  FilterSelect,
+  FilterText,
```
These are not necessarily limited to use with filtering. Maybe:

`InputText`
`InputSelect`
`InputDateRange`

galelis replied:
Liked it. I'll update.

dannyrb commented on platform/ui/index.js:
```diff
@@ -10,6 +10,10 @@ export {
+  FilterDateRange,
+  FilterSelect,
+  FilterText,
+  FilterWrapper,
```
Not sure what this does yet. Will come back if naming feels off.

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+  label,
+  isSortable,
+  isBeingSorted,
+  sortDirection,
```
Should this sort stuff be composable? Part of some HOC magic?

dannyrb replied:
I think we can cut this down to:

`isSortable` and `sortDirection`
It's being sorted if `isSortable` is true and `sortDirection` is `ascending` or `descending` (or whatever their corresponding values are)

galelis replied:
Updated :)

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+  sortDirection,
+  onLabelClick,
+  inputValue,
+  inputProps,
```
This implies a leaky abstraction. Can we just expose the props directly that we want to pass through so the user doesn't have to look up docs for `DateRange`?

dannyrb replied:
That would means these would be renamed to `value`. With `inputProps` expanded based on the props we want to pass through.

galelis replied:
I've updated the code to work only with the required properties on each input component.

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+  isSortable,
+  isBeingSorted,
+  sortDirection,
+  onLabelClick,
```
Should this be changed to `onSortUpdate` or something similar?

dannyrb replied:
Maybe `onLabelClick` does make more sense 🤔 
LMK your thoughts.

galelis replied:
I have added LabelClick because used can do anything he wanted in case any click on label accours.

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+  }),
+  inputProps: PropTypes.object,
+  onChange: PropTypes.func,
+};
```
When possible, default to `required` with no default state. It's less variations we have to support and account for. For example, `isSortable` should simply be required. As should label, and a few others.

galelis replied:
Done. make sense.

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+    startDate: PropTypes.string,
+    endDate: PropTypes.string,
+  }),
+  inputProps: PropTypes.object,
```
This is not specific enough. I know we're going to make these more explicit, but almost always want `shape` if we're going to support an object.

galelis replied:
I've removed it from the input components. Passing only the needed props

dannyrb commented on platform/ui/src/components/FilterDateRange/FilterDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+  onLabelClick: PropTypes.func,
+  value: PropTypes.shape({
+    startDate: PropTypes.string,
+    endDate: PropTypes.string,
```
Are these string? Do they have to be in a specific format? Inline comment to clarify?

galelis replied:
PropTypes updated. Comment added.

dannyrb commented on platform/ui/src/components/FilterSelect/FilterSelect.jsx:
```diff
@@ -0,0 +1,66 @@
+  onLabelClick,
+  inputValue,
+  inputProps,
+  onChange,
```
Same notes on props for this component.

galelis replied:
Updated the code.

dannyrb commented on platform/ui/src/components/FilterSelect/FilterSelect.jsx:
```diff
@@ -0,0 +1,66 @@
+      onLabelClick={onLabelClick}
+    >
+      <Select
+        {...inputProps}
```
Never spread props 🙏

dannyrb commented on platform/ui/src/components/FilterSelect/FilterSelect.jsx:
```diff
@@ -0,0 +1,66 @@
+  isBeingSorted: PropTypes.bool,
+  sortDirection: PropTypes.number,
+  onLabelClick: PropTypes.func,
+  value: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
```
array of? shape?

galelis replied:
Updated :)

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
```
Maybe renamed "FormGroup"
Similar semantics of Bootstrap where you wrap label + input + extra
Not necessarily limited to filtering

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+  label: PropTypes.string,
+  isSortable: PropTypes.bool,
+  isBeingSorted: PropTypes.bool,
+  sortDirection: PropTypes.number,
```
Maybe "one of" if we only support three values

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+        {label}
+        {isSortable && <Icon {...iconProps} />}
+      </span>
+      <span>{children && children}</span>
```
`children && children` would have the same effect as `children`?

dannyrb replied:
But why inside a `<span>`?

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+  sortDirection: PropTypes.number,
+  onLabelClick: PropTypes.func,
+  className: PropTypes.string,
+  children: PropTypes.node,
```
100% should be required.

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+  const iconProps = {
+    name: isBeingSorted ? sortIconMap[sortDirection] : 'sorting',
+    className: classnames(
+      'mx-2 w-2',
```
In general, the "outer element" of a React Component should not have margin.
The consumer of the component can set that, and this limits the places where this component can be used (with limited benefit)

In Viewer "view", this would be set on the outside of each element, and that would be cool beans

galelis replied:
This is actually the iconProps.className that is used only inside the "FilterWrapper".  
I agree with you and I think we are fine on this one.

dannyrb commented on platform/ui/src/components/FilterWrapper/FilterWrapper.jsx:
```diff
@@ -0,0 +1,67 @@
+    name: isBeingSorted ? sortIconMap[sortDirection] : 'sorting',
+    className: classnames(
+      'mx-2 w-2',
+      isBeingSorted ? 'text-custom-aquaBright' : 'text-custom-blue'
```
CC: @rodrigolabs, have we not addressed colors yet? Any blockers?

rodrigolabs replied:
I created a specific ticket for color naming.

dannyrb commented on platform/ui/src/components/Select/Select.jsx:
```diff
@@ -55,7 +52,7 @@ Select.propTypes = {
   placeholder: PropTypes.string,
   noOptionsMessage: PropTypes.func,
-  value: PropTypes.string,
+  value: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
```
Same refactoring energy here

dannyrb commented on platform/ui/src/views/StudyList/StudyList.js:
```diff
@@ -11,55 +11,57 @@ const filtersMeta = [
     name: 'instances',
     displayName: 'Instances',
-    inputType: 'none',
+    inputType: 'None',
```
Make sure PropTypes reflects available options for `inputType`

dannyrb commented on platform/ui/src/views/StudyList/StudyList.js:
```diff
@@ -11,55 +11,57 @@ const filtersMeta = [
+        { value: 'seg', label: 'seg' },
+        { value: 'ct', label: 'ct' },
+        { value: 'mr', label: 'mr' },
+        { value: 'sr', label: 'sr' },
```
Caps

dannyrb commented on platform/ui/src/components/FilterText/FilterText.jsx:
```diff
@@ -0,0 +1,60 @@
+  onChange: PropTypes.func,
+};
+
+export default FilterText;
```
Same notes as other components

dannyrb commented on platform/ui/src/views/StudyList/components/StudyListFilter.js:
```diff
@@ -241,19 +143,24 @@ const StudyListFilter = ({
-                      >
-                        {renderInput(inputType, name, { selectOptions })}
-                      </FilterLabel>
+                      {FilterInputComponent && (
```
Just mark as required?

galelis replied:
This is for cases like "Instances" where we don't have an input but we have the label and sorting. 
Shoul we treat that in a different way? Like creating a component for that expecifically?

dannyrb replied:
Good question. Is the "just a label" a component? Maybe I skipped over where/how you're rendering it.

galelis replied:
As it could have sorting on it, I have used the same stucture but I don't add an input for it.

dannyrb commented on platform/ui/src/views/StudyList/components/StudyListFilter.js:
```diff
@@ -275,12 +182,32 @@ const StudyListFilter = ({
       name: PropTypes.string,
       dsplayName: PropTypes.string,
-      inputType: PropTypes.oneOf(['text', 'select', 'date-range', 'none']),
+      inputType: PropTypes.oneOf(['Text', 'Select', 'DateRange', 'None']),
```
Ah. Perfect. Thanks!

dannyrb commented on platform/ui/src/views/StudyList/components/StudyListFilter.js:
```diff
@@ -275,12 +182,32 @@ const StudyListFilter = ({
+    resultsPerPage: 25,
+  },
+  numOfStudies: 90,
+};
```
On the fence about defaults here. Less pressing as this is higher up; might be worth forcing above layer to provide values so its easier to see at a glance where everything is coming from.

We should have a dedicated page for just `StudyListFilter`

rodrigolabs commented on platform/ui/src/components/InputDateRange/InputDateRange.jsx:
```diff
@@ -0,0 +1,56 @@
+          startDate={startDate}
+          endDate={endDate}
+          onChange={({ startDate, endDate }) => {
+            if (onChange) {
```
As `onChange` is required, maybe we don't need this condition here.

rodrigolabs commented on platform/ui/src/components/InputDateRange/InputDateRange.jsx:
```diff
@@ -0,0 +1,56 @@
+
+InputDateRange.propTypes = {
+  label: PropTypes.string.isRequired,
+  isSortable: PropTypes.bool.isRequired,
```
`isSortable` is really necessary? it's only used to define if the sort icon should be displayed or not. Maybe we can use the `sortDirection` itself to define it.

- If `sortDirection` has any value, the field is automatically sortable, right?
- `sortDirection` can be "not required" -- field is not sortable if `sortDirection` is empty or  has`none` as value.

What do you think? Open to ideas.

dannyrb replied:
I'm okay with either option, so long as its documented.

I'm trying to think about a parent setting all as sortable or not. Does it make more sense to set this per input, or to clear sort direction? What would you expect as a consumer of this library?

galelis replied:
I have 2 thoughts here. 
>Maybe we can use the sortDirection itself to define it.

`isSortable + sortDirection` could be a single `sorting` with values as `['none', 'ascending', 'descending', 'hidden']` but the way its now, is more intuitive to understand the direction and if we show or not. 
> Does it make more sense to set this per input, or to clear sort direction?

rodrigolabs commented on platform/ui/src/components/InputSelect/InputSelect.jsx:
```diff
@@ -0,0 +1,84 @@
+        className="mt-2"
+        options={options}
+        value={value}
+        isMulti={true}
```
`<Select />` will always be `multi`?

dannyrb replied:
I'm okay with it being locked down and potentially renaming this `MultiSelect`

Again, less options is better, unless we need that level of configuration. It's less to test, and these are UI states we won't be able to verify the accuracy of as they're not encountered in the application.

If we find ourselves needing a non-multi-select, then we can revisit the API here?

rodrigolabs replied:
Perfect! I agree. Let's keep as it is for now!

galelis replied:
I thoguht the same as Danny, as he is always suggesting to lower the options if we can, as our input field is a multiSelect for now, I decided to move with it locking this option. 

But I definetely liked the idea of renaming to InputMultiSelect.

rodrigolabs commented on platform/ui/src/components/InputText/InputText.jsx:
```diff
@@ -0,0 +1,48 @@
+
+InputText.propTypes = {
+  label: PropTypes.string.isRequired,
+  isSortable: PropTypes.bool.isRequired,
```
Same thing about `isSortable`

rodrigolabs commented on platform/ui/src/components/Select/Select.jsx:
```diff
@@ -1,61 +1,115 @@
 import './Select.css';
 
+const MultiValueContainer = ({ data }) => {
+  return <span>{`${data.label}, `}</span>;
```
Is it possible to have a comma separating the values only if there are two or more values?
To avoid this:

![image](https://user-images.githubusercontent.com/1905961/77666214-71338700-6f5f-11ea-8103-275d86bc5f2c.png)

dannyrb replied:
Array.split(', ') works well for this; then just display the string

galelis replied:
I will try to work with that. 
Select allows me to override some of its internal components, I was overriding only the MultiSelectValue which only allows me to touch 1 by 1 of the selected items. 

I will work on this to avoid displaying the coma on the last item.

galelis replied:
Fixed.

rodrigolabs commented on platform/ui/src/components/Select/Select.jsx:
```diff
@@ -1,61 +1,115 @@
+          checked={props.isSelected}
+          className="w-6 h-6"
+          onChange={e => null}
+        />{' '}
```
Was this "blank space" `{' '}` intentional?

rodrigolabs replied:
Checkbox + Text is not aligned:

<img src="https://user-images.githubusercontent.com/1905961/77666804-29f9c600-6f60-11ea-83b2-5bdae0cef74a.png" width="300" />

dannyrb replied:
I think autofix inserts the blank space to preserve line breaks that would not persist with current formatting rules.

galelis replied:
Fixed alingment and also the extra `{' '}`.

rodrigolabs commented on platform/ui/src/components/Select/Select.jsx:
```diff
@@ -1,61 +1,115 @@
       className={classnames(
         className,
-        'flex flex-col flex-1 mt-2 customSelect__wrapper'
+        'flex flex-col flex-1 customSelect__wrapper'
```
Maybe we could define a `min-width` for this component, here's how it's being displayed in the docz page for Select:

![image](https://user-images.githubusercontent.com/1905961/77667445-f8cdc580-6f60-11ea-93eb-0fb85797345d.png)

I know it depends on the parent width, but to avoid break lines like this, we could just set a `min-width`.

What do you think?

galelis replied:
I haven't touch the react-select css for that. I can do it thought. 
Ideally it works like this so it can expand and always show all selected itens. 
I don't think setting a min with will solve this. cause it will depends on the size of the options as well. 

We would need to think on how to deal with bigger inputs, as we did for the InputMultiSelect, when we wrap the content and add ellipses. But this was an especific scenario.

rodrigolabs replied:
Ok, let's only add a bigger width in the doc page as suggested

rodrigolabs commented on platform/ui/src/views/StudyList/components/StudyListFilter.js:
```diff
@@ -108,61 +59,43 @@ const StudyListFilter = ({
+        isSortable={_isSortable}
+        sortDirection={_sortDirection}
+        onLabelClick={onLabelClick}
+        {...inputProps}
```
`inputProps` is currently used for the Select options, right? As we should avoid spreading props, isn't it better to explicitly pass this value to an `options` prop?

galelis replied:
The code will be a little bit bigger now, but I removed this spreading porps.

rodrigolabs commented on platform/ui/src/components/Select/Select.mdx:
```diff
@@ -22,10 +22,10 @@ import { Select } from '@ohif/ui';
     <div className="w-20">
```
About the Select `min-width`, let's change this width and then we're good!

```suggestion
    <div className="w-56">
```

rodrigolabs commented:
Only two changes:

- Remove the condition for `onChange` in `InputDateRange.jsx`
- Increase the width of the `div` wrapping the Select component in the Select doc page

And then we're good to merge!

Awesome work! 👏🏻

galelis commented:
@dannyrb @rodrigolabs 
All requested changes done. 
- Renamed InputSelect to InputMultiSelect
- Removed the condition for `onChange` in `InputDateRange.jsx`
- Updated Select docz page adding bigger width on select
